README.md

# EpiSoon

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-blue.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R build
status](https://github.com/epiforecasts/EpiSoon/workflows/R-CMD-check/badge.svg)](https://github.com/epiforecasts/EpiSoon)
[![Codecov test
coverage](https://codecov.io/gh/epiforecasts/EpiSoon/branch/main/graph/badge.svg)](https://app.codecov.io/gh/epiforecasts/EpiSoon?branch=main)
[![MIT
license](https://img.shields.io/badge/License-MIT-blue.svg)](https://github.com/epiforecasts/EpiSoon/blob/main/LICENSE.md/)
[![GitHub
contributors](https://img.shields.io/github/contributors/epiforecasts/EpiSoon)](https://github.com/epiforecasts/EpiSoon/graphs/contributors)
[![universe](https://epiforecasts.r-universe.dev/badges/EpiSoon)](http://epiforecasts.r-universe.dev/ui/#package:EpiSoon)
[![DOI](https://zenodo.org/badge/248311916.svg)](https://zenodo.org/badge/latestdoi/248311916)

This package provides tooling to forecast the time-varying reproduction
number and use this to forecast reported case counts via a branching
process. It supports a range of time series modelling packages including
`bsts`, `forecast`, and `fable`. It also supports ensembles via `stackr`
and `forecastHyrbid`. Forecasts can be assessed by iteractively fitting
and then using proper scoring rules (via `scoringutils` and
`scoringRules`) to compare to both observed case counts and estimated
reproduction numbers.

## Installation

Install the stable development version of the package with:

``` r
install.packages("EpiSoon", repos = "https://epiforecasts.r-universe.dev")
```

Install the unstable development version of the package with (few users
should need to do this):

``` r
remotes::install_github("epiforecasts/EpiSoon")
```

## Quick start

  - Load packages (`bsts` and `fable` for models, `ggplot2` for
    plotting, and `cowplot` for theming)

<!-- end list -->

``` r
library(EpiSoon)
library(bsts)
library(fable)
library(future)
library(cowplot)
library(dplyr)
```

  - Set up example data (using `EpiSoon::example_obs_rts` and
    `EpiSoon::example_obs_cases` as starting data sets). When generating
    timeseries with `EpiNow` use `get_timeseries` to extract the
    required data.

<!-- end list -->

``` r
obs_rts <- EpiSoon::example_obs_rts %>%
  dplyr::mutate(timeseries = "Region 1") %>%
  dplyr::bind_rows(EpiSoon::example_obs_rts %>%
    dplyr::mutate(timeseries = "Region 2"))

obs_cases <- EpiSoon::example_obs_cases %>%
  dplyr::mutate(timeseries = "Region 1") %>%
  dplyr::bind_rows(EpiSoon::example_obs_cases %>%
    dplyr::mutate(timeseries = "Region 2"))
```

  - Define the list of models to be compared.

<!-- end list -->

``` r
models <- list(
  "AR 3" =
    function(...) {
      EpiSoon::bsts_model(
        model =
          function(ss, y) {
            bsts::AddAr(ss, y = y, lags = 3)
          }, ...
      )
    },
  "Semi-local linear trend" =
    function(...) {
      EpiSoon::bsts_model(
        model =
          function(ss, y) {
            bsts::AddSemilocalLinearTrend(ss, y = y)
          }, ...
      )
    },
  "ARIMA" =
    function(...) {
      EpiSoon::fable_model(model = fable::ARIMA(y ~ time), ...)
    }
)
```

  - Compare models across timeseries (change the `future::plan` to do
    this in parallel).

<!-- end list -->

``` r
future::plan("sequential")

## Compare models
forecasts <- EpiSoon::compare_timeseries(obs_rts, obs_cases, models,
  horizon = 7, samples = 10,
  serial_interval = EpiSoon::example_serial_interval
)
#> Warning: There were 40 warnings in `dplyr::mutate()`.
#> The first warning was:
#> ℹ In argument: `eval = furrr::future_pmap(...)`.
#> Caused by warning:
#> ! Unknown or uninitialised column: `.distribution`.
#> ℹ Run `dplyr::last_dplyr_warnings()` to see the 39 remaining warnings.

forecasts
#> $forecast_rts
#> # A tibble: 511 × 12
#>    timeseries model foreca…¹ date       horizon median  mean     sd bottom lower
#>    <chr>      <chr> <chr>    <date>       <int>  <dbl> <dbl>  <dbl>  <dbl> <dbl>
#>  1 Region 1   AR 3  2020-03… 2020-03-05       1   2.26  2.25 0.0504   2.17  2.25
#>  2 Region 1   AR 3  2020-03… 2020-03-06       2   2.21  2.19 0.0729   2.09  2.19
#>  3 Region 1   AR 3  2020-03… 2020-03-07       3   2.15  2.12 0.0957   1.95  2.13
#>  4 Region 1   AR 3  2020-03… 2020-03-08       4   2.07  2.05 0.128    1.86  2.06
#>  5 Region 1   AR 3  2020-03… 2020-03-09       5   2.00  1.98 0.165    1.69  2.00
#>  6 Region 1   AR 3  2020-03… 2020-03-10       6   1.95  1.93 0.173    1.61  1.93
#>  7 Region 1   AR 3  2020-03… 2020-03-11       7   1.90  1.86 0.199    1.52  1.86
#>  8 Region 1   AR 3  2020-03… 2020-03-07       1   2.11  2.11 0.0389   2.02  2.10
#>  9 Region 1   AR 3  2020-03… 2020-03-08       2   2.04  2.03 0.0449   1.94  2.02
#> 10 Region 1   AR 3  2020-03… 2020-03-09       3   1.97  1.94 0.0713   1.80  1.96
#> # … with 501 more rows, 2 more variables: upper <dbl>, top <dbl>, and
#> #   abbreviated variable name ¹​forecast_date
#> 
#> $rt_scores
#> # A tibble: 399 × 12
#>    timeseries model forec…¹ date       horizon    mad  bias   dss   crps log_s…²
#>    <chr>      <chr> <chr>   <date>       <int>  <dbl> <dbl> <dbl>  <dbl>   <dbl>
#>  1 Region 1   AR 3  2020-0… 2020-03-05       1 0.0514  -0.2 -5.98 0.0110  -1.96 
#>  2 Region 1   AR 3  2020-0… 2020-03-06       2 0.0630   0.4 -5.34 0.0176  -1.55 
#>  3 Region 1   AR 3  2020-0… 2020-03-07       3 0.0879   0   -4.72 0.0232  -1.31 
#>  4 Region 1   AR 3  2020-0… 2020-03-08       4 0.129   -0.2 -4.01 0.0345  -0.961
#>  5 Region 1   AR 3  2020-0… 2020-03-09       5 0.165   -0.4 -3.44 0.0514  -0.764
#>  6 Region 1   AR 3  2020-0… 2020-03-10       6 0.170   -0.4 -3.25 0.0578  -0.713
#>  7 Region 1   AR 3  2020-0… 2020-03-11       7 0.186   -0.6 -2.67 0.0902  -0.483
#>  8 Region 1   AR 3  2020-0… 2020-03-07       1 0.0283  -0.8 -5.53 0.0207  -2.05 
#>  9 Region 1   AR 3  2020-0… 2020-03-08       2 0.0273  -1   -3.32 0.0506  -0.712
#> 10 Region 1   AR 3  2020-0… 2020-03-09       3 0.0398  -1   -2.13 0.0864  -0.162
#> # … with 389 more rows, 2 more variables: ae_median <dbl>, se_mean <dbl>, and
#> #   abbreviated variable names ¹​forecast_date, ²​log_score
#> 
#> $forecast_cases
#> # A tibble: 399 × 12
#>    timeseries model forecas…¹ date       horizon median  mean    sd bottom lower
#>    <chr>      <chr> <chr>     <date>       <int>  <dbl> <dbl> <dbl>  <dbl> <dbl>
#>  1 Region 1   AR 3  2020-03-… 2020-03-05       1   68.5  67.5  5.91     59    61
#>  2 Region 1   AR 3  2020-03-… 2020-03-06       2   81.5  80.5 11.8      61    72
#>  3 Region 1   AR 3  2020-03-… 2020-03-07       3   97    94.4 15.7      68    94
#>  4 Region 1   AR 3  2020-03-… 2020-03-08       4  114   116.  13.9     102   102
#>  5 Region 1   AR 3  2020-03-… 2020-03-09       5  139   131.  24.0      98   132
#>  6 Region 1   AR 3  2020-03-… 2020-03-10       6  150.  156.  23.9     117   141
#>  7 Region 1   AR 3  2020-03-… 2020-03-11       7  186.  182.  41.1     122   155
#>  8 Region 1   AR 3  2020-03-… 2020-03-07       1   94.5  92.7  8.30     80    94
#>  9 Region 1   AR 3  2020-03-… 2020-03-08       2  108   107.   6.34     95   104
#> 10 Region 1   AR 3  2020-03-… 2020-03-09       3  122.  121.  13.8      94   115
#> # … with 389 more rows, 2 more variables: upper <dbl>, top <dbl>, and
#> #   abbreviated variable name ¹​forecast_date
#> 
#> $case_scores
#> # A tibble: 399 × 12
#>    timeseries model sample forecast…¹ date       horizon   mad  bias   dss  crps
#>    <chr>      <chr> <chr>  <chr>      <date>       <int> <dbl> <dbl> <dbl> <dbl>
#>  1 Region 1   AR 3  1      2020-03-04 2020-03-05       1  6.67   0.4  4.09  2.75
#>  2 Region 1   AR 3  1      2020-03-04 2020-03-06       2 11.1    0.4  5.28  4.77
#>  3 Region 1   AR 3  1      2020-03-04 2020-03-07       3 13.3    0.4  5.58  6.3 
#>  4 Region 1   AR 3  1      2020-03-04 2020-03-08       4 14.1    0.6  6.30  6.85
#>  5 Region 1   AR 3  1      2020-03-04 2020-03-09       5 29.7    0.2  6.70 11.1 
#>  6 Region 1   AR 3  1      2020-03-04 2020-03-10       6 14.1    0.6  6.68  7.7 
#>  7 Region 1   AR 3  1      2020-03-04 2020-03-11       7 42.3    0.2  7.47 13.1 
#>  8 Region 1   AR 3  1      2020-03-06 2020-03-07       1  8.90   0.4  4.48  3.65
#>  9 Region 1   AR 3  1      2020-03-06 2020-03-08       2  5.19   0.6  4.23  3.68
#> 10 Region 1   AR 3  1      2020-03-06 2020-03-09       3 13.3    0.2  5.26  4.1 
#> # … with 389 more rows, 2 more variables: ae_median <dbl>, se_mean <dbl>, and
#> #   abbreviated variable name ¹​forecast_date
```

  - Plot an evaluation of Rt forecasts using iterative fitting.

<!-- end list -->

``` r
EpiSoon::plot_forecast_evaluation(forecasts$forecast_rts, obs_rts, c(7)) +
  ggplot2::facet_grid(model ~ timeseries) +
  cowplot::panel_border()
```

<img src="man/figures/unnamed-chunk-8-1.png" width="60%" />

  - Plot an evaluation of case forecasts using iterative fitting

<!-- end list -->

``` r
EpiSoon::plot_forecast_evaluation(forecasts$forecast_cases, obs_cases, c(7)) +
  ggplot2::facet_grid(model ~ timeseries, scales = "free") +
  cowplot::panel_border()
```

<img src="man/figures/unnamed-chunk-9-1.png" width="60%" />

  - Summarise the forecasts by model scored against observed cases

<!-- end list -->

``` r
EpiSoon::summarise_scores(forecasts$case_scores)
#> # A tibble: 12 × 9
#>    score     model             bottom lower median    mean  upper    top      sd
#>    <chr>     <chr>              <dbl> <dbl>  <dbl>   <dbl>  <dbl>  <dbl>   <dbl>
#>  1 ae_median AR 3               0.85   9     24.5  5.69e+1   91   2.27e2 6.28e+1
#>  2 ae_median Semi-local linea…  0.613  8.5   24.2  5.11e+1   92.4 1.88e2 5.86e+1
#>  3 bias      AR 3              -1      0      0.6  4.47e-1    1   1   e0 6.13e-1
#>  4 bias      Semi-local linea… -1      0.2    0.6  4.81e-1    1   1   e0 5.90e-1
#>  5 crps      AR 3               2.94   6.57  17.8  4.69e+1   79.3 2.00e2 5.59e+1
#>  6 crps      Semi-local linea…  2.80   6.82  15.6  4.18e+1   69.5 1.60e2 5.12e+1
#>  7 dss       AR 3               4.53   6.09   8.33 1.29e+1   15.9 4.67e1 1.09e+1
#>  8 dss       Semi-local linea…  4.56   6.20   7.74 1.28e+1   12.7 5.07e1 1.35e+1
#>  9 mad       AR 3               7.93  13.3   21.5  2.61e+1   34.1 8.51e1 1.76e+1
#> 10 mad       Semi-local linea…  6.10  12.8   19.3  2.46e+1   31.1 6.58e1 1.70e+1
#> 11 se_mean   AR 3               1.15  81    784    7.29e+3 8855.  4.72e4 1.56e+4
#> 12 se_mean   Semi-local linea…  0.524 79.2  595.   6.26e+3 8663.  4.16e4 1.31e+4
```

## Contributing

File an issue [here](https://github.com/epiforecasts/EpiSoon/issues) if
you have identified an issue with the package. Please note that due to
operational constraints priority will be given to users informing
government policy or offering methodological insights. We welcome all
contributions, in particular those that improve the approach or the
robustness of the code base.


NEWS.md
# EpiSoon 0.3.1

This is a maintenance release which updates the package CI and use of functions depreciated since the last release in 2020. It also includes a range of other changes that were missed from previous releases.

Thanks to @nikosbosse, @andreamra, @tuxette, @sbfnk, and @seabbs for code contributions towards this release.

* Updated all GitHub Actions
* Updated `tidyverse` code usage to account for depreciation.
* Updated `pkgdown` site to use `pkgdown` 2.0.0.
* Updated all uses of `size` to `linewidth` to account for depreciation in `ggplot2` line geoms.

# EpiSoon 0.3.0

* Reviewed and updated all tooling and examples
* Added documentation
* Bugs squashed.

# EpiSoon 0.2.0

* Added a `NEWS.md` file to track changes to the package.
* Added dev support for multiple samples to `evaluate_model` and all higher order functions. Note that this user is expected to manage this themselves for lower level functions. 
* Renamed models to reflect forecasting and not fitting.
* Updated all docs to reflect changes.
* Added case prediction and scoring framework


R/plot.R
#' Plot a Forecast
#'
#' @param forecast A dataframe with summarised forecasts as produced
#' by `summarise_forecast` or `summarise_case_forecast` .
#' @param observations A dataframe of observations containing the
#' following variables:
#' - either `rt` or `cases`
#' - and `date`.
#' @param horizon_cutoff Numeric, defaults to NULL. Forecast horizon to plot up
#' to.
#' @param obs_cutoff_at_forecast Logical defaults to `TRUE`. Should
#' observations only be shown
#' up to the date of the forecast.
#' @importFrom dplyr filter
#' @importFrom ggplot2 ggplot aes geom_line geom_ribbon scale_x_date labs
#' @importFrom cowplot theme_cowplot
#' @importFrom rlang has_name
#' @return A `ggplot2` object
#' @export
#'
#' @examples
#' \dontrun{
#' ## Forecast an Rt sample
#' samples <- forecast_rt(EpiSoon::example_obs_rts[1:10, ],
#'   model = function(...) {
#'     EpiSoon::bsts_model(
#'       model =
#'         function(ss, y) {
#'           bsts::AddSemilocalLinearTrend(ss, y = y)
#'         }, ...
#'     )
#'   },
#'   horizon = 21, samples = 10
#' )
#'
#' ## Summarise forecast
#' summarised_forecast <- summarise_forecast(samples)
#'
#' ## Plot forecast_cases
#' plot_forecast(summarised_forecast, EpiSoon::example_obs_rts)
#'
#' ## Forecast a case sample
#' pred_cases <- forecast_cases(EpiSoon::example_obs_cases, samples,
#'   serial_interval = EpiSoon::example_serial_interval
#' )
#'
#' summarised_case_forecast <- summarise_case_forecast(pred_cases)
#'
#' plot_forecast(summarised_case_forecast, EpiSoon::example_obs_cases)
#' }
plot_forecast <- function(forecast = NULL,
                          observations = NULL,
                          horizon_cutoff = NULL,
                          obs_cutoff_at_forecast = TRUE) {
  if (obs_cutoff_at_forecast) {
    observations <- observations %>%
      dplyr::filter(date < min(forecast$date))
  }

  if (!is.null(horizon_cutoff)) {
    forecast <- forecast %>%
      dplyr::filter(horizon <= horizon_cutoff)
  }

  if ("cases" %in% colnames(observations)) {
    case_plot <- TRUE
    observations <- observations %>%
      dplyr::mutate(y = cases)
  } else {
    case_plot <- FALSE
    observations <- observations %>%
      dplyr::mutate(y = rt)
  }

  plot <- ggplot2::ggplot(forecast, ggplot2::aes(x = date)) +
    ggplot2::geom_line(ggplot2::aes(y = bottom), alpha = 0.5) +
    ggplot2::geom_line(ggplot2::aes(y = top), alpha = 0.5) +
    ggplot2::geom_ribbon(ggplot2::aes(ymin = bottom, ymax = top), alpha = 0.1) +
    ggplot2::geom_ribbon(
      ggplot2::aes(ymin = lower, ymax = upper), alpha = 0.2
    ) +
    ggplot2::geom_point(
      data = observations,
      ggplot2::aes(y = y), size = 1.1,
      alpha = ifelse(rlang::has_name(observations, "sample"),
        max(1 / max(observations$sample, na.rm = TRUE), 0.01),
        0.7
      )
    ) +
    cowplot::theme_cowplot() +
    ggplot2::scale_x_date(date_breaks = "1 week", date_labels = "%b %d") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))

  if (case_plot) {
    plot <- plot + ggplot2::labs(x = "Date", y = "New daily cases")
  } else {
    plot <- plot + ggplot2::labs(x = "Date", y = "Effective Reproduction no.")
  }

  return(plot)
}



#' Plot a Forecast
#'
#' @param forecasts A dataframe as produced by `forecast_rt`
#' or `forecast_cases`
#' @param horizon_to_plot Numeric vector, the forecast horizon to plot.
#'
#' @inheritParams plot_forecast
#' @importFrom dplyr filter
#' @importFrom ggplot2 ggplot aes geom_line geom_ribbon scale_x_date labs
#' @importFrom cowplot theme_cowplot
#' @return A `ggplot2` object
#' @export
#'
#' @examples
#' \dontrun{
#' ## Evaluate a model
#' forecast_eval <- evaluate_model(EpiSoon::example_obs_rts,
#'   EpiSoon::example_obs_cases,
#'   model = function(...) {
#'     EpiSoon::bsts_model(
#'       model =
#'         function(ss, y) {
#'           bsts::AddSemilocalLinearTrend(ss, y = y)
#'         }, ...
#'     )
#'   },
#'   serial_interval = EpiSoon::example_serial_interval,
#'   horizon = 7, samples = 10
#' )
#'
#' ## Plot Rt forecast
#' plot_forecast_evaluation(forecast_eval$forecast_rts,
#'   EpiSoon::example_obs_rts,
#'   horizon_to_plot = 7
#' )
#'
#'
#' ## Plot case forecast
#' plot_forecast_evaluation(forecast_eval$forecast_cases,
#'   EpiSoon::example_obs_cases,
#'   horizon_to_plot = 7
#' )
#' }
plot_forecast_evaluation <- function(forecasts = NULL,
                                     observations = NULL,
                                     horizon_to_plot = 1) {
  forecasts <- forecasts %>%
    dplyr::filter(horizon %in% horizon_to_plot)

  plot <- plot_forecast(forecasts,
    observations,
    obs_cutoff_at_forecast = FALSE
  )

  return(plot)
}


#' Plot forecast scores
#'
#' @return A dataframe of summarised scores in a tidy format.
#' @export
plot_scores <- function() {
  ##  Some thought required here as to what the best - most general purpose
  ## scoring plot would be.
}



#' Summary plots to compare timeseries and forecast models
#'
#' @param compare_timeseries_output A named list of dataframes produced by
#' \code{compare_timeseries}
#' @param type Type(s) of summary plot to be produced for Rt and case
#' observations for each model in n \code{compare_timeseries_output}.
#' \code{"summary_score"} provides a plot of model fit scores for
#' the 0-7 and 8-14 day horizons. If desired, a subset of scores can be
#' specified using the \code{score} argument. \code{"horizon_score"}
#' provides a plot of scores (CRPS, Calibration, Sharpness, Median, IQR,
#' CI, Bias) across horizons. \code{"region_score"} provides a plot of
#' scores (CRPS, Calibration, Sharpness, Bias, Median, IQR) by region for
#' the 0-7 and 8-14 day horizons.
#' @param score (Optional) One or more of
#' \code{c("Bias","CRPS","Sharpness","Calibration","Median","IQR","CI")} when
#' \code{type="summary_score"} is used.
#'
#' @return A named list of `ggplot2` objects
#' @export
#' @importFrom dplyr mutate bind_rows filter group_by ungroup recode_factor
#' pick
#' @importFrom cowplot plot_grid theme_cowplot panel_border
#' @importFrom purrr map_dfr
#' @importFrom lubridate days
#' @importFrom ggplot2 ggplot labs aes geom_point position_dodge geom_linerange
#' @importFrom ggplot2 theme facet_wrap coord_flip geom_line
#' @importFrom ggplot2 scale_x_continuous scale_fill_viridis_d facet_grid
#' @importFrom ggplot2 scale_y_log10 scale_color_viridis_d aes_string
#'
#' @examples
#' \dontrun{
#' obs_rts <- EpiSoon::example_obs_rts %>%
#'   dplyr::mutate(timeseries = "Region 1") %>%
#'   dplyr::bind_rows(EpiSoon::example_obs_rts %>%
#'     dplyr::mutate(timeseries = "Region 2"))
#'
#' obs_cases <- EpiSoon::example_obs_cases %>%
#'   dplyr::mutate(timeseries = "Region 1") %>%
#'   dplyr::bind_rows(EpiSoon::example_obs_cases %>%
#'     dplyr::mutate(timeseries = "Region 2"))
#'
#' models <- list(
#'   "AR 3" = function(...) {
#'     EpiSoon::bsts_model(model = function(ss, y) {
#'       bsts::AddAr(ss, y = y, lags = 3)
#'     }, ...)
#'   },
#'   "Semi-local linear trend" = function(...) {
#'     EpiSoon::bsts_model(model = function(ss, y) {
#'       bsts::AddSemilocalLinearTrend(ss, y = y)
#'     }, ...)
#'   }
#' )
#'
#' forecast_eval <-
#'   compare_timeseries(obs_rts, obs_cases, models,
#'     horizon = 10, samples = 10,
#'     serial_interval = EpiSoon::example_serial_interval
#'   )
#'
#' ## Produce all plots
#' plot_compare_timeseries(forecast_eval)
#'
#' ## Produce subsets of plots
#' plot_compare_timeseries(forecast_eval, type = "summary_score")
#' plot_compare_timeseries(forecast_eval,
#'   type = "summary_score",
#'   score = "Bias"
#' )
#' plot_compare_timeseries(forecast_eval, type = "horizon_score")
#' plot_compare_timeseries(forecast_eval, type = "region_score")
#' plot_compare_timeseries(forecast_eval,
#'   type = c("horizon_score", "region_score")
#' )
#' }
plot_compare_timeseries <- function(compare_timeseries_output,
                                    type = c(
                                      "summary_score",
                                      "horizon_score",
                                      "region_score"
                                    ),
                                    score = c(
                                      "Bias",
                                      "CRPS",
                                      "Dispersion",
                                      "AE (median)",
                                      "SE (mean)"
                                    )) {
  ## Prepare plotting output
  plot_output <- list()

  ## Pull in results
  rt_scores <- compare_timeseries_output$rt_scores
  case_scores <- compare_timeseries_output$case_scores

  ## Identify maximum available horizon
  max_horizon <- max(rt_scores$horizon)

  if ("summary_score" %in% type) {
    sum_scores <-
      summarise_scores_recent_all(rt_scores, recent = 7) %>%
      dplyr::mutate(type = "Rt") %>%
      dplyr::bind_rows(summarise_scores_recent_all(case_scores,
        recent = 7
      ) %>%
        dplyr::mutate(type = "Cases")) %>%
      adjust_score(group_var = c("horizon", "type", "data")) %>%
      dplyr::mutate(type = type %>%
        factor(levels = c("Rt", "Cases")))

    for (score_choice in score) {
      plot_output[[paste0(tolower(score_choice), "_summary_plot")]] <-
        summary_plot(sum_scores, score_choice)
    }
  }
  if ("horizon_score" %in% type) {
    plot_rt_horizon_score <-
      EpiSoon::summarise_scores(rt_scores, "horizon") %>%
      adjust_score("horizon") %>%
      make_horizon_score_plot(label = "Rt")

    plot_cases_horizon_score <-
      EpiSoon::summarise_scores(case_scores, "horizon") %>%
      adjust_score("horizon") %>%
      make_horizon_score_plot(label = "Cases")

    ##  plot_horizon_score
    plot_output[["horizon_score_plot"]] <-
      cowplot::plot_grid(plot_rt_horizon_score,
        plot_cases_horizon_score,
        ncol = 1
      )
  }
  if ("region_score" %in% type) {
    ## Summarised scores
    scores_0_7 <- list(rt = rt_scores, case = case_scores) %>%
      purrr::map_dfr(
        ~ dplyr::filter(., horizon <= 7) %>%
          EpiSoon::summarise_scores("timeseries"),
        .id = "type"
      )

    plot_output[["region_score_7_plot"]] <-
      scores_0_7 %>%
      dplyr::filter(!score %in% "ci") %>%
      make_region_score_plot(title = "Horizon: 0 to 7")

    if (max_horizon > 7) {
      scores_8_plus <- list(rt = rt_scores, case = case_scores) %>%
        purrr::map_dfr(
          ~ dplyr::filter(., horizon > 7) %>%
            EpiSoon::summarise_scores("timeseries"),
          .id = "type"
        )
      plot_output[["region_score_7_up_plot"]] <-
        scores_8_plus %>%
        dplyr::filter(!score %in% "ci") %>%
        make_region_score_plot(title = "Horizon: 7+") +
        ggplot2::labs(x = "", y = "")
    }
  }
  return(plot_output)
}

## Internal unexported functions for plotting/summary ------------------

adjust_score <- function(df, group_var) {
  df_update <- df %>%
    dplyr::group_by(score, dplyr::pick({{ group_var }})) %>%
    dplyr::mutate(upper_min = 10 * min(upper)) %>%
    dplyr::ungroup() # %>%
  df_update <-
    df_update[which(df_update$upper <= df_update$upper_min |
      df_update$score %in% "bias"), ] %>%
    dplyr::ungroup() %>%
    dplyr::filter(!score %in% c("log_score", "dss")) %>%
    dplyr::mutate(score = score %>%
      factor(levels = c(
        "crps", "bias",
        "ae_median", "mad", "se_mean"
      )) %>%
      dplyr::recode_factor(
        "crps" = "CRPS",
        "bias" = "Bias",
        "ae_median" = "AE (median)",
        "mad" = "Dispersion",
        "se_mean" = "SE (mean)"
      ))
  return(df_update)
}

## Summarise scores
summarise_scores_by_horizon <- function(scores) {
  score_7 <- scores %>%
    dplyr::filter(horizon <= 7) %>%
    EpiSoon::summarise_scores() %>%
    dplyr::mutate(horizon = "0 -- 7")

  score_14 <- scores %>%
    dplyr::filter(horizon <= 14, horizon > 7) %>%
    EpiSoon::summarise_scores() %>%
    dplyr::mutate(horizon = "8 -- 14")

  scores <- score_7 %>%
    dplyr::bind_rows(score_14) %>%
    # dplyr::bind_rows(score_14_plus) %>%
    dplyr::mutate(horizon = horizon %>%
      factor(levels = c("0 -- 7", "8 -- 14")))

  return(scores)
}

## Summarise by horizon for all and recent data only
summarise_scores_recent_all <- function(scores, recent = NULL) {
  all_scores <- scores %>%
    summarise_scores_by_horizon()

  recent_scores <- scores %>%
    dplyr::filter(date >= max(date) - lubridate::days(recent)) %>%
    summarise_scores_by_horizon()

  sum_scores <- all_scores %>%
    dplyr::mutate(data = "All") %>%
    dplyr::bind_rows(recent_scores %>%
      dplyr::mutate(data = paste0(
        "Most recent ",
        recent, " days"
      )))
  return(sum_scores)
}

## Region score plot
plot_region_score <- function(scores, label = NULL) {
  scores %>%
    adjust_score("timeseries") %>%
    ggplot2::ggplot(ggplot2::aes(
      x = timeseries, y = median,
      col = model
    )) +
    ggplot2::geom_point(
      size = 1.1,
      position = ggplot2::position_dodge(width = 1)
    ) +
    ggplot2::geom_point(ggplot2::aes(y = median),
      shape = 2, size = 1.1,
      position = ggplot2::position_dodge(width = 1)
    ) +
    ggplot2::geom_linerange(ggplot2::aes(ymin = lower, ymax = upper),
      alpha = 0.4, linewidth = 1.1,
      position =
        ggplot2::position_dodge(width = 1)
    ) +
    ggplot2::scale_color_viridis_d(option = "cividis") +
    cowplot::theme_cowplot() +
    ggplot2::theme(legend.position = "bottom") +
    ggplot2::labs(
      col = "Model", y = "Score value", x = "Region",
      tag = label
    ) +
    ggplot2::facet_wrap(~score, scales = "free_x") +
    ggplot2::coord_flip()
}

## Internal plot
plot_internal <- function(df, label = NULL) {
  plot <- df %>%
    ggplot2::ggplot(ggplot2::aes(
      x = horizon, y = mean, col = model,
      group = model
    )) +
    ggplot2::geom_line(linewidth = 1.2, alpha = 0.6) +
    ggplot2::geom_point(size = 2) +
    ggplot2::geom_point(ggplot2::aes(y = median), shape = 2, size = 2) +
    ggplot2::geom_linerange(ggplot2::aes(ymin = lower, ymax = upper),
      alpha = 0.4, linewidth = 1.5,
      position =
        ggplot2::position_dodge(width = 3)
    ) +
    ggplot2::scale_color_viridis_d(option = "cividis") +
    ggplot2::scale_x_continuous(breaks = unique(df$horizon)) +
    cowplot::theme_cowplot() +
    ggplot2::theme(legend.position = "bottom") +
    ggplot2::labs(col = "Model", y = "", x = "Horizon", tag = label) +
    ggplot2::facet_wrap(~score, scales = "free_y")

  return(plot)
}

# Generate summary plots
summary_plot <- function(scores, target_score) {
  scores %>%
    dplyr::filter(score %in% target_score) %>%
    ggplot2::ggplot(
      ggplot2::aes_string(x = "model", y = "mean", col = "data")
    ) +
    ggplot2::geom_point(
      size = 2,
      position = ggplot2::position_dodge(width = 1)
    ) +
    ggplot2::geom_point(ggplot2::aes(y = median),
      shape = 2, size = 2,
      position = ggplot2::position_dodge(width = 1)
    ) +
    ggplot2::geom_linerange(ggplot2::aes(ymin = lower, ymax = upper),
      alpha = 0.4, linewidth = 1.5,
      position =
        ggplot2::position_dodge(width = 1)
    ) +
    ggplot2::scale_fill_viridis_d(
      option = "cividis", begin = 0.2,
      end = 0.6
    ) +
    cowplot::theme_cowplot() +
    ggplot2::theme(legend.position = "bottom") +
    ggplot2::coord_flip() +
    ggplot2::facet_grid(horizon ~ type, scales = "free_x") +
    cowplot::panel_border() +
    ggplot2::labs(x = "Model", y = target_score, col = "Data")
}

## Summarise scores across time horizons
make_horizon_score_plot <- function(df, label = NULL) {
  df <- df %>%
    dplyr::filter(horizon %% 3 == 0, horizon <= 15)

  score_plot <- df %>%
    dplyr::filter(!score %in% "Bias") %>%
    dplyr::filter(mean > 0, median > 0, lower > 0, upper > 0) %>%
    plot_internal(label = label) +
    ggplot2::scale_y_log10()

  bias_plot <- df %>%
    dplyr::filter(score %in% "Bias") %>%
    plot_internal(label = "") +
    ggplot2::theme(legend.position = "none")

  plot <- cowplot::plot_grid(score_plot, bias_plot,
    nrow = 1, rel_widths = c(3, 1)
  )

  return(plot)
}

## Summarise scores across regions
make_region_score_plot <- function(df, title = NULL) {
  plot_region_score_rt <- df[which(df$type %in% "rt"), ] %>%
    # dplyr::filter(type %in% "rt") %>%
    plot_region_score(label = "Rt") +
    ggplot2::labs(title = title, x = "", y = "")

  plot_region_score_cases <- df[which(df$type %in% "case"), ] %>%
    # dplyr::filter(type %in% "case") %>%
    plot_region_score(label = "Cases") +
    ggplot2::theme(legend.position = "none")

  plot_region_score_new <- cowplot::plot_grid(plot_region_score_rt,
    plot_region_score_cases,
    nrow = 1
  )
  return(plot_region_score_new)
}
